### PR TITLE
skip rows which are added after cursor created

### PR DESCRIPTION
**Pull request by kaijianding**
The checking maxId behavior is weird, We can get dims array from cursor's currEntry in IndexerDimensionSelector, but choosing to ignore the dims if dims are beyond maxId which is actually too old if data arrives after the cursor is made and cursor can retrieve really new data when the new data is inside the cursor's interval.
But think about this case, a group by query may produces aggregator values with empty string dimensions, while actually these aggregator values should bound to some non-empty dimensions.

--update
the reason why maxId check is needed is explained here https://github.com/druid-io/druid/pull/4052#issuecomment-286779784
but still we can do it in a better way as described at https://github.com/druid-io/druid/pull/4052#issuecomment-288976998

The behavior test is already in IncrementalIndexStorageAdapterTest.testCursoringAndIndexUpdationInterleaving()

### CODE REVIEW

**Comment by gianm** [comment]
I believe the purpose of this maxId check is to prevent a selector from returning a value that is outside the bound of what `getValueCardinality()` returns before the caller starts walking through the cursor. This would violate the contract of DimensionSelector.

I guess you could work around your original issue by filtering out empty strings. Would that work or do you think a different solution is needed?

Btw: I just raised #4052 to fix a slight inconsistency here.

**Comment by kaijianding** [comment]
#4052 looks good, I will backport #4052 and try it in my environment to see if my origin issue is gone.

**Comment by gianm** [comment]
I think it probably won't fix your original issue by itself, unless you also have a filter like `not(selector(dim, ""))` to filter out empties. If you do that, your original issue should be fixed, I thin.

**Comment by kaijianding** [comment]
My original use case is like this:
I cached the DImensionSelector, and then go through the cursor. When cursor comes to the new data, the cursor.isDone() returns false, then I use the cached DImensionSelector to getRow().get(0) -- I know I have a dimension which contains only 1 not-null value, so I can call get(0) without any check, and then "Index (0) is greater than or equal to list size (0)" exception is raised.
I finally identify that it is due to the id>=maxId logic.

I think if cursor.isDone() returns true when the cursor arrives to the new data, then there is no chance to call getRow().get(0) and should avoid the exception. 
If valueMatcher handles the maxId correctly, the cursor.isDone() should also behaves correctly.
Will look into this part of code to do something like #4052 -- I'm still using 0.9.2 and #4052 may not be able to be applied directly.

**Comment by gianm** [comment]
@kaijianding, see the thread in https://github.com/druid-io/druid/pull/4052#issuecomment-288976998 for a possible solution to your problem that maintains the DimensionSelector contract. I'm not able to work on that patch right now, so if you are able to then go for it. Otherwise I may come back to it later.

**Comment by kaijianding** [comment]
OK, I will modify this PR to do that.@gianm

**Comment by kaijianding** [comment]
It's ready for review, @gianm @leventov

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndexStorageAdapter.java [hunk 1]
```diff
@@ -276,7 +278,15 @@ public void advance()
                 while (baseIter.hasNext()) {
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndexStorageAdapter.java]: `advanceUninterruptibly()` should be also changed.

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndexStorageAdapter.java [hunk 2]
```diff
@@ -362,6 +387,10 @@ public void reset()
                 done = !foundMatched && (emptyRange || !baseIter.hasNext());
               }
 
+              private boolean beyondMaxRowIndex(Integer rowIndex) {
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndexStorageAdapter.java]: Use primitive argument

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndexStorageAdapter.java [hunk 3]
```diff
@@ -350,7 +368,14 @@ public void reset()
 
                 boolean foundMatched = false;
                 while (baseIter.hasNext()) {
-                  currEntry.set(baseIter.next());
+                  Map.Entry<IncrementalIndex.TimeAndDims, Integer> entry = baseIter.next();
+                  // ignore rows whose rowIndex is beyond the maxRowIndex
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndexStorageAdapter.java]: Move this comment to `beyondMaxRowIndex()`

File: processing/src/main/java/io/druid/segment/StringDimensionIndexer.java [hunk 4]
```diff
@@ -415,23 +413,15 @@ public IndexedInts getRow()
...
-              row = IntArrays.EMPTY_ARRAY;
-              rowSize = 0;
-            }
+            row = new int[] {nullId};
+            rowSize = 1;
           }
         }
 
         if (row == null && indices != null && indices.length > 0) {
           row = new int[indices.length];
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: Don't need to copy array to a new place,
```java
row = indices;
rowSize = indices.length;
```

File: processing/src/main/java/io/druid/segment/StringDimensionIndexer.java [hunk 5]
```diff
@@ -415,23 +413,15 @@ public IndexedInts getRow()
...
           if (nullId > -1) {
-            if (nullId < maxId) {
-              row = new int[] {nullId};
-              rowSize = 1;
-            } else {
-              // Choose to use ArrayBasedIndexedInts later, instead of EmptyIndexedInts, for monomorphism
-              row = IntArrays.EMPTY_ARRAY;
-              rowSize = 0;
-            }
+            row = new int[] {nullId};
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: `ArrayBasedIndexedInts` could be memoized here, instead of creating new array and new ArrayBasedIndexedInts every time.

File: processing/src/main/java/io/druid/segment/StringDimensionIndexer.java [hunk 6]
```diff
@@ -415,23 +413,15 @@ public IndexedInts getRow()
...
 
         if (row == null && indices != null && indices.length > 0) {
           row = new int[indices.length];
           for (int id : indices) {
-            if (id < maxId) {
-              row[rowSize++] = id;
-            }
+            row[rowSize++] = id;
           }
         }
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: If `indices == null` (i. e. `dimIndex >= dims.length`) and `nullId` not set yet (-1), there will be an NPE. Is that sane? I don't know. Maybe @gianm knows.

**Comment by leventov** [comment]
Please also change the title of the PR. Originally instead of completely changing the contents of the PR it was better to close this PR and create a new one.

File: processing/src/main/java/io/druid/segment/StringDimensionIndexer.java (hunk 5 above)
Comment (gianm) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: This is the same behavior as before, fine by me unless the author is motivated to tweak it.

File: processing/src/main/java/io/druid/segment/StringDimensionIndexer.java [hunk 7]
```diff
@@ -509,7 +499,7 @@ public boolean matches()
       @Override
       public int getValueCardinality()
       {
-        return maxId;
+        return getCardinality();
```
Comment (gianm) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: This should still be `maxId`. The selector should act as if it's reading a snapshot of the column that existed at the time the selector was created. So, `getValueCardinality` should be frozen to the snapshot.

Comment (gianm) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: I think it'd be good to put something like this in `lookupName` too, as a sanity check:

```java
if (id >= maxId) {
  throw new ISE("id[%d] >= maxId[%d]", id, maxId);
}
```

File: processing/src/main/java/io/druid/segment/StringDimensionIndexer.java (hunk 6 above)
Comment (gianm) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: This block is protected by an `indices != null` check, so it won't be a problem.

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndexStorageAdapter.java (hunk 3 above)
Comment (gianm) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndexStorageAdapter.java]: I assume you mean from all three places. And, I agree.

File: processing/src/main/java/io/druid/segment/StringDimensionIndexer.java (hunk 6 above)
Comment (leventov) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: @gianm, no, there will be an NPE in line 438 (actually inside `ArrayBasedIndexedInts.of()`), because `row` will be `null`. I wasn't able to comment on it, because it "doesn't belong to the PR".

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndexStorageAdapter.java (hunk 2 above)
Comment (kaijianding) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndexStorageAdapter.java]: The entry.getValue() is Integer, isn't better to use Integer to avoid boxing/unboxing?
>                   Map.Entry<IncrementalIndex.TimeAndDims, Integer> entry = baseIter.next();
>                   if (beyondMaxRowIndex(entry.getValue())) {
>                     continue;
>                   }

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndexStorageAdapter.java (hunk 1 above)
Comment (kaijianding) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndexStorageAdapter.java]: already changed to pass the CI

File: processing/src/main/java/io/druid/segment/StringDimensionIndexer.java (hunk 5 above)
Comment (kaijianding) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: Done memoizing the nullId int array

File: processing/src/main/java/io/druid/segment/StringDimensionIndexer.java (hunk 6 above)
Comment (kaijianding) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: @leventov you are right, there maybe NPE. I change the code to make sure row=IntArrays.EMPTY_ARRAY if nullId <=-1, then row is always not null.

**Comment by kaijianding** [comment]
comments addressed and ut added @gianm @leventov

File: processing/src/main/java/io/druid/segment/StringDimensionIndexer.java (hunk 6 above)
Comment (gianm) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: Sorry, I misunderstood the problem you were talking about because of the line number thing. It should still be fine though, since row being null would lead to `return ArrayBasedIndexedInts.of(null, 0)`, which returns `ArrayBasedIndexedInts.EMPTY`.

Comment (gianm) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: Although that might just be "lucky". I'm not sure if ArrayBasedIndexInts.of on `null` is really meant to do that by contract, versus throw NPE. It doesn't have a javadoc explaining the contract.

File: processing/src/test/java/io/druid/segment/incremental/IncrementalIndexStorageAdapterTest.java [hunk 8]
```diff
@@ -458,4 +458,157 @@ public Object apply(Cursor cursor)
...
+      );
+    }
+
+    final StorageAdapter sa = new IncrementalIndexStorageAdapter(index);
+
+    Sequence<Cursor> cursors = sa.makeCursors(
+        null, new Interval(timestamp - 60_000, timestamp + 60_000), VirtualColumns.EMPTY, Granularities.ALL, false
+    );
+
+    Sequences.toList(
```
Comment (gianm) [inline processing/src/test/java/io/druid/segment/incremental/IncrementalIndexStorageAdapterTest.java]: Please include something here to make sure that the code actually gets executed. It's possible that `cursors` is empty and none of these assertions are run.

Either checking the return value of Sequences.toList, or incrementing an AtomicInteger inside the map fn and checking it after Sequences.toList returns, should be enough.

File: processing/src/test/java/io/druid/segment/incremental/IncrementalIndexStorageAdapterTest.java [hunk 9]
```diff
@@ -458,4 +458,157 @@ public Object apply(Cursor cursor)
...
+                }
+
+                DimensionSelector dimSelectorC = cursor.makeDimensionSelector(
+                    new DefaultDimensionSpec(
+                        "billy",
+                        "billy"
+                    )
+                );
+
+                DimensionSelector dimSelectorD = cursor.makeDimensionSelector(
```
Comment (gianm) [inline processing/src/test/java/io/druid/segment/incremental/IncrementalIndexStorageAdapterTest.java]: These names are getting confusing, how about calling dimSelect C `dimSelector1C` and dimSelector D `dimSelector2A` respectively (number referring to the "billy" column and letter referring to its order amongst dim selectors for that column)?

File: processing/src/test/java/io/druid/segment/incremental/IncrementalIndexStorageAdapterTest.java [hunk 10]
```diff
@@ -458,4 +458,157 @@ public Object apply(Cursor cursor)
...
+
+                DimensionSelector dimSelectorE = cursor.makeDimensionSelector(
+                    new DefaultDimensionSpec(
+                        "billy3",
+                        "billy3"
+                    )
+                );
+
+                // and then, cursoring continues in the other thread
+                while (!cursor.isDone()) {
```
Comment (gianm) [inline processing/src/test/java/io/druid/segment/incremental/IncrementalIndexStorageAdapterTest.java]: Are any of these assertions actually firing? I would think that this cursor should be empty and so `cursor.isDone()` will return true immediately -- isn't that what this patch is trying to achieve?

Comment (gianm) [inline processing/src/test/java/io/druid/segment/incremental/IncrementalIndexStorageAdapterTest.java]: Please double-check this test case, as well as the existing `testCursoringAndIndexUpdationInterleaving` case, which I think is also kind of invalid now.

File: processing/src/main/java/io/druid/segment/StringDimensionIndexer.java [hunk 11]
```diff
@@ -412,27 +415,26 @@ public IndexedInts getRow()
 
         int[] row = null;
         int rowSize = 0;
+        // typically, currEntry's rowIndex is smaller than the row's rowIndex in which this dim first appears
```
Comment (gianm) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: What does this comment mean? It doesn't really make sense to me.

**Comment by gianm** [comment]
Main code looks good to me but I had some comments about the tests.

File: processing/src/main/java/io/druid/segment/StringDimensionIndexer.java (hunk 11 above)
Comment (kaijianding) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: add comment to explain in which situation indices can be null

File: processing/src/test/java/io/druid/segment/incremental/IncrementalIndexStorageAdapterTest.java (hunk 8 above)
Comment (kaijianding) [inline processing/src/test/java/io/druid/segment/incremental/IncrementalIndexStorageAdapterTest.java]: Added CountDownLatch to make sure cursors are not empty

File: processing/src/test/java/io/druid/segment/incremental/IncrementalIndexStorageAdapterTest.java (hunk 9 above)
Comment (kaijianding) [inline processing/src/test/java/io/druid/segment/incremental/IncrementalIndexStorageAdapterTest.java]: Modified

File: processing/src/test/java/io/druid/segment/incremental/IncrementalIndexStorageAdapterTest.java (hunk 10 above)
Comment (kaijianding) [inline processing/src/test/java/io/druid/segment/incremental/IncrementalIndexStorageAdapterTest.java]: because cant't make dimensionSelectorA from empty cursor from empty index, so I just starting from a non-empty index. Then `cursor.isDone()` is not always true

**Comment by kaijianding** [comment]
ut comments addressed @gianm

File: processing/src/test/java/io/druid/segment/incremental/IncrementalIndexStorageAdapterTest.java (hunk 10 above)
Comment (gianm) [inline processing/src/test/java/io/druid/segment/incremental/IncrementalIndexStorageAdapterTest.java]: Oh, I see, you added rows to the index before calling makeCursors.

File: processing/src/test/java/io/druid/segment/incremental/IncrementalIndexStorageAdapterTest.java [hunk 12]
```diff
@@ -450,12 +452,173 @@ public Object apply(Cursor cursor)
...
+                  }
+                  IndexedInts rowD = dimSelector2D.getRow();
+                  // no null id, so should get empty dims array
+                  Assert.assertEquals(0, rowD.size());
+                  IndexedInts rowE = dimSelector3E.getRow();
+                  Assert.assertEquals(1, rowE.size());
+                  // the null id
+                  Assert.assertEquals(0, rowE.get(0));
+                  cursor.advance();
+                  rowNumInCursor.countDown();
```
Comment (gianm) [inline processing/src/test/java/io/druid/segment/incremental/IncrementalIndexStorageAdapterTest.java]: Please use AtomicInteger + increment for this. Otherwise, we can't properly measure the number of rows that get returned by the cursor. (countDown on a latch with value 0 is ignored)

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndexStorageAdapter.java (hunk 2 above)
Comment (gianm) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndexStorageAdapter.java]: rowIndex has to get unboxed anyway for comparison, but using `compareTo` additionally forces boxing of maxRowIndex. So accepting a primitive argument at least avoids the boxing of maxRowIndex.

File: processing/src/test/java/io/druid/segment/incremental/IncrementalIndexStorageAdapterTest.java (hunk 12 above)
Comment (kaijianding) [inline processing/src/test/java/io/druid/segment/incremental/IncrementalIndexStorageAdapterTest.java]: done

**Comment by kaijianding** [comment]
this check at IncrementalIndexStorageAdapter line 293
```java
                if (!filterMatcher.matches()) {
                  done = true;
                }
```
I think `done` should always be true at this place, I found this code was written from the very beginning, could you help to verify I understood it correctly? @gianm

File: processing/src/main/java/io/druid/segment/StringDimensionIndexer.java (hunk 6 above)
Comment (leventov) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: It was a lucky coincidence. `ArrayBasedIndexedInts.of(null)` throws NPE.

**Comment by gianm** [comment]
Hmm, I don't see a reason to check `!filterMatcher.matches()` there, as it should always be true. I think it's fine to make the change you did.

**Comment by kaijianding** [comment]
@gianm @leventov do you have more comments?

**Comment by fjy** [comment]
👍

**Comment by leventov** [comment]
Will take a final look tomorrow.

File: processing/src/main/java/io/druid/segment/StringDimensionIndexer.java [hunk 13]
```diff
@@ -412,27 +415,27 @@ public IndexedInts getRow()
...
-              row = new int[] {nullId};
-              rowSize = 1;
-            } else {
-              // Choose to use ArrayBasedIndexedInts later, instead of EmptyIndexedInts, for monomorphism
-              row = IntArrays.EMPTY_ARRAY;
-              rowSize = 0;
+            if (nullIdIntArray == null) {
+              nullIdIntArray = new int[] {nullId};
             }
+            row = nullIdIntArray;
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: There is a race, the single element of `nullIdIntArray`, created in another thread, may not be visible in this thread yet. You can resolve this race by caching the whole `ArrayBasedIndexedInts` object, not just array.

File: processing/src/main/java/io/druid/segment/StringDimensionIndexer.java (hunk 6 above)
Comment (leventov) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: @kaijianding could you add explicit nonNull checks in ArrayBasedIndexedInts factory methods, and add ternary `row != null ? ArrayBasedIndexedInts.of(row, rowSize) : ArrayBasedIndexedInts.empty();` in this getRow() method?

File: processing/src/main/java/io/druid/segment/StringDimensionIndexer.java (hunk 13 above)
Comment (gianm) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: @leventov do you know for sure if this is going to be a problem?

As far as I know, dimension selectors don't get passed between threads in an unsafe way.

Comment (gianm) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: They should be either used by the thread that created them, or (like in groupBy v2) used under a synchronization.

Comment (leventov) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: @gianm it is unspecified. And also it's anyway a good idea to cache ArrayBasedIndexedInts here, to additionally cut garbage creation.

Comment (gianm) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: I think we could safely specify that selectors are not expected to be thread safe. They are usually wrappers over a cursor which is not thread safe. But sure, changing to ArrayBasedIndexedInts won't hurt anything.

Comment (gianm) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: I'm good with this patch either way. It's not worse than before in terms of garbage, and it should be safe to make thread-unsafe DimensionSelectors (see also: SingleScanTimeDimSelector). @kaijianding I think it's up to you if you want to change it or not. Please let us know.

Comment (leventov) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: I approved too.
